### PR TITLE
kubelet: memory manager: fix preferred topology hints calculation

### LINKED ISSUE #104682: Memory Manager allows Guaranteed QoS Pod with hugepages requested is exactly equal to the left over Hugepages
#### What happened:
Setup:
1. 3 Master nodes and 2 Worker Nodes. 
2. Kubernetes Version: v1.22.0-rc.0+5c2f7cd
3. Configuring Memory Manager on Kubernetes Version (v1.22.0-rc.0+5c2f7cd)  and Nodes configured with Hugepages of size 2M on 2 Worker nodes. 
4. Each Worker node having 2 Numa Nodes (numa node0 and numa node1)
5. Number of Hugepages of size 2M are 10 on each Numa node 

sh-4.4# cat /sys/devices/system/node/node0/hugepages/hugepages-2048kB/nr_hugepages      
10
sh-4.4# cat /sys/devices/system/node/node1/hugepages/hugepages-2048kB/nr_hugepages 
10

After configuring Memory , where Hugepages are 20M (2M * 10)  on each Numa node. we create QoS pod1 pod which consumes 24Mi Hugepages (of size 2M) 
Pod Spec:

```
apiVersion: v1
kind: Pod
metadata:
  name: pod1
spec:
  containers:
  - name: example
    image: fedora:latest
    command:
    - sleep
    - inf
    volumeMounts:
    - mountPath: /hugepages-2Mi
      name: hugepage-2mi
    resources:
      limits:
        hugepages-2Mi: 24Mi
        memory: "24Mi"
        cpu: "2"
      requests:
        hugepages-2Mi: 24Mi
        memory: "24Mi"
        cpu: "2"
  volumes:
  - name: hugepage-2mi
    emptyDir:
      medium: HugePages-2Mi
```
List pods

```
$ oc get pods
NAME   READY   STATUS    RESTARTS   AGE   IP             NODE                              NOMINATED NODE   READINESS GATES
pod1   1/1     Running   0          13s   10.128.2.109   worker1.example.org           <none>           <none>
```
Memory Manager state file:
```
{"policyName":"Static","machineState":{"0":{"numberOfAssignments":2,"memoryMap":{"hugepages-1Gi":{"total":0,"systemReserved":0,"allocatable":0,"reserved":0,"free":0},"hugepages-2Mi":{"total":20971520,"systemReserved":0,"allocatable":20971520,"reserved":20971520,"free":0},"memory":{"total":270146174976,"systemReserved":1153433600,"allocatable":268971769856,"reserved":104857600,"free":268866912256}},"cells":[0,1]},"1":{"numberOfAssignments":2,"memoryMap":{"hugepages-1Gi":{"total":0,"systemReserved":0,"allocatable":0,"reserved":0,"free":0},"hugepages-2Mi":{"total":20971520,"systemReserved":0,"allocatable":20971520,"reserved":4194304,"free":16777216},"memory":{"total":270531874816,"systemReserved":0,"allocatable":270510903296,"reserved":0,"free":270510903296}},"cells":[0,1]}},"entries":{"bbf8fd78-3c9d-4924-b4a2-450caaca6da3":{"example":[{"numaAffinity":[0,1],"type":"hugepages-2Mi","size":25165824},{"numaAffinity":[0,1],"type":"memory","size":104857600}]}},"checksum":1759502831}
```
Now create a QoS Guaranteed Pod2 requesting 16Mi . On the same worker node. 

Pod spec:
```
apiVersion: v1
kind: Pod
metadata:
  name: pod2
spec:
  containers:
  - name: example
    image: fedora:latest
    command:
    - sleep
    - inf
    volumeMounts:
    - mountPath: /hugepages-2Mi
      name: hugepage-2mi
    resources:
      limits:
        hugepages-2Mi: "16Mi"
        memory: "100Mi"
        cpu: "2"
      requests:
        hugepages-2Mi: "16Mi"
        memory: "100Mi"
        cpu: "2"
  nodeSelector:
    kubernetes.io/hostname: "worker1.example.org "
  volumes:
  - name: hugepage-2mi
    emptyDir:
      medium: HugePages-2Mi
```
List pods:

```
NAME   READY   STATUS    RESTARTS   AGE     IP             NODE                              NOMINATED NODE   READINESS GATES
pod1   1/1     Running   0          4m39s   10.128.2.109   worker1.example.org              <none>                     <none>
pod2   1/1     Running   0          11s     10.128.2.111   worker1.example.org                 <none>                     <none>
```
Memory Manager state file after pod2 is deployed.

```
{"policyName":"Static","machineState":{"0":{"numberOfAssignments":4,"memoryMap":{"hugepages-1Gi":{"total":0,"systemReserved":0,"allocatable":0,"reserved":0,"free":0},"hugepages-2Mi":{"total":20971520,"systemReserved":0,"allocatable":20971520,"reserved":20971520,"free":0},"memory":{"total":270146174976,"systemReserved":1153433600,"allocatable":268971769856,"reserved":209715200,"free":268762054656}},"cells":[0,1]},"1":{"numberOfAssignments":4,"memoryMap":{"hugepages-1Gi":{"total":0,"systemReserved":0,"allocatable":0,"reserved":0,"free":0},"hugepages-2Mi":{"total":20971520,"systemReserved":0,"allocatable":20971520,"reserved":20971520,"free":0},"memory":{"total":270531874816,"systemReserved":0,"allocatable":270510903296,"reserved":0,"free":270510903296}},"cells":[0,1]}},"entries":{"90a01c04-cfc4-401b-bdb8-85667384f002":{"example":[{"numaAffinity":[0,1],"type":"hugepages-2Mi","size":16777216},{"numaAffinity":[0,1],"type":"memory","size":104857600}]},"bbf8fd78-3c9d-4924-b4a2-450caaca6da3":{"example":[{"numaAffinity":[0,1],"type":"hugepages-2Mi","size":25165824},{"numaAffinity":[0,1],"type":"memory","size":104857600}]}},"checksum":3930981289}
```

#### What you expected to happen:
Pod2 should be rejected. 

#### How to reproduce it (as minimally and precisely as possible):
Steps provided above
#### Environment:
- Kubernetes version (use `kubectl version`): v1.22.0-rc.0+5c2f7cd
- Cloud provider or hardware configuration: Openshift
- OS (e.g: `cat /etc/os-release`):Red Hat Enterprise Linux CoreOS release 4.9
- Kernel (e.g. `uname -a`): Linux helix02.lab.eng.tlv2.redhat.com 4.18.0-305.12.1.rt7.84.el8_4.x86_64 #1 SMP PREEMPT_RT Thu Jul 29 14:18:12 EDT 2021 x86_64 x86_64 x86_64 GNU/Linux

Kubelet configuration 
```
spec:
  kubeletConfig:
    apiVersion: kubelet.config.k8s.io/v1beta1
    authentication:
      anonymous: {}
      webhook:
        cacheTTL: 0s
      x509: {}
    authorization:
      webhook:
        cacheAuthorizedTTL: 0s
        cacheUnauthorizedTTL: 0s
    cpuManagerPolicy: static
    cpuManagerReconcilePeriod: 5s
    evictionHard:
      memory.available: 100Mi
    evictionPressureTransitionPeriod: 0s
    fileCheckFrequency: 0s
    httpCheckFrequency: 0s
    imageMinimumGCAge: 0s
    kind: KubeletConfiguration
    kubeReserved:
      cpu: 1000m
      memory: 500Mi
    logging: {}
    memoryManagerPolicy: Static
    nodeStatusReportFrequency: 0s
    nodeStatusUpdateFrequency: 0s
    reservedMemory:
    - limits:
        memory: 1100Mi
      numaNode: 0
    reservedSystemCPUs: 0-4,40-44
    runtimeRequestTimeout: 0s
    shutdownGracePeriod: 0s
    shutdownGracePeriodCriticalPods: 0s
    streamingConnectionIdleTimeout: 0s
    syncFrequency: 0s
    systemReserved:
      cpu: 1000m
      memory: 500Mi
    topologyManagerPolicy: restricted
    volumeStatsAggPeriod: 0s
  machineConfigPoolSelector:
    matchLabels:
      machineconfiguration.openshift.io/role: worker-cnf

```

### PR DESCRIPTION
**Pull request by cynepco3hahue**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Prevent starting pods with resources satisfied by a single NUMA node on multiple NUMA nodes.
The code returned before it updated the minimal amount of NUMA nodes that can satisfy the container
requests.

#### Which issue(s) this PR fixes:

Fixes #104682

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

Signed-off-by: Artyom Lukianov <alukiano@redhat.com>

### CODE REVIEW

**Review by matthyx — APPROVED** [review]
/lgtm

**Comment by cynepco3hahue** [comment]
The memory manager lane fails because of the lack of a Dynamic Kubelet feature gate.

**Comment by 249043822** [comment]
This change makes caculating resources first to ensure set minAffinitySize correctly, looks nice.

**Comment by klueska** [comment]
Can you explain this change a bit better? It's not immediately obvious from the code change what the old semantics were vs. what the desired new semantics should. Which makes it hard to verify that the new semantics are correct.

**Comment by cynepco3hahue** [comment]
> Can you explain this change a bit better? It's not immediately obvious from the code change what the old semantics were vs. what the desired new semantics should be. Which makes it hard to verify that the new semantics are correct.

Sure, will try to explain it better. The problem is that during `bitmask.IterateBitMasks` when we are checking all possible permutations of NUMA nodes it is possible that we will return from the method for the current permutation before updating the `minAffinitySize`. And once we are calculating if the hins are preferred or not, we can get the wrong result because of the wrong `minAffinitySize`. The PR only re-organizes a little bit of the code that we will sure that we always will update the `minAffinitySize`.

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 1]
```diff
@@ -438,28 +438,10 @@ func (p *staticPolicy) calculateHints(machineState state.NUMANodeMap, pod *v1.Po
 		maskBits := mask.GetBits()
 		singleNUMAHint := len(maskBits) == 1
 
-		// the node already in group with another node, it can not be used for the single NUMA node allocation
-		if singleNUMAHint && len(machineState[maskBits[0]].Cells) > 1 {
-			return
-		}
-
```
Comment (klueska) [inline pkg/kubelet/cm/memorymanager/policy_static.go:445]: I see. So basically you move this call below the loop (so we don't break out early), and then repurpose the loop to *only* update `minAffinitySize`(if possible).

Then, if you make it through this loop (with a new `minAffinity` or one equal to a previously calculated `minAffinity`), only then do you go through the process of calculating the set of hints using (basically) the same logic you did before.

I didn't verify in this review that the existing logic is necessarily correct (I assume we did that in a previous PR), but the minor refactoring here seems makes sense.

**Pull request merged**